### PR TITLE
CFG - While Node Visitor

### PR DESCRIPTION
**Pull request by kavins14**
Added `while` visitor method as well as the corresponding tests.

### CODE REVIEW

**Comment by kavins14** [comment]
I feel that checking the `expected edges` in the `while` visitor tests can be done with fewer lines of code by comparing the block ids (instead of the statements). However, I decided to write them this way is because it is much easier to understand. Also, checking for `expected blocks` is somewhat implied when checking `expected edges` but I still decided to include them because it would allow us to debug block-related errors more easily.

File: python_ta/cfg/graph.py [hunk 1]
```diff
@@ -31,9 +31,12 @@ def link_or_merge(self, source: CFGBlock, target: CFGBlock) -> None:
         An "empty" node for this purpose is when source has no statements.
         """
         if source.statements == []:
-            for edge in source.predecessors:
-                edge.target = target
-                target.predecessors.append(edge)
+            if source.predecessors == []:   # implies source == empty start node
```
Comment (david-yz-liu) [inline python_ta/cfg/graph.py]: Use `if source is self.start` here.

File: tests/test_cfg/test_while.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        [["print(n + 1)"], ["print(n)"]],
+        [["print(n)"], []]
+    ]
+    assert expected_edges == _extract_edges(cfg)
+
+
+def test_while_with_if() -> None:
+    src="""
+    while n > 10:
+        if (n > 20):
```
Comment (david-yz-liu) [inline tests/test_cfg/test_while.py]: Remove the parentheses around the if condition (throughout).

File: tests/test_cfg/test_while.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        [["n > 10"], ["print(n)"]],
+        [["print(n)"], ["n > 10"]],
+        [["n > 10"], ["print(n + 1)"]],
+        [["print(n + 1)"], []]
+    ]
+    assert expected_edges == _extract_edges(cfg)
+
+
+def test_simple_while_with_surrounding_blocks() -> None:
+    src="""
```
Comment (david-yz-liu) [inline tests/test_cfg/test_while.py]: Put spaces around the `=` (throughout this file).

File: python_ta/cfg/graph.py [hunk 4]
```diff
@@ -31,9 +31,12 @@ def link_or_merge(self, source: CFGBlock, target: CFGBlock) -> None:
         An "empty" node for this purpose is when source has no statements.
         """
         if source.statements == []:
-            for edge in source.predecessors:
-                edge.target = target
-                target.predecessors.append(edge)
+            if source is self.start:
+                self.start = target
```
Comment (dhslim) [inline python_ta/cfg/graph.py]: add self.block_count = 1  here?

Comment (dhslim) [inline python_ta/cfg/graph.py]: add self.block_count = 1  here?

Comment (dhslim) [inline python_ta/cfg/graph.py]: self.block_count = 1 included here?

Comment (kavins14) [reply python_ta/cfg/graph.py]: Good catch, the issue with including `self.block_count = 1` is that we might have multiple instances of `CFGBlock` with the same id. This also applies if we include `self.block_count -= 1` instead (why I think would be more appropriate). A possible fix is that we use the built-in `id()` function to set the id of the blocks. Then we would be able to adjust the `block_count` attribute in the `link_or_merge` method.

Comment (david-yz-liu) [reply python_ta/cfg/graph.py]: Yes, this is a good question. Right now `block_count` is used as an "autoincrement" to ensure the block ids are unique, and we'll take the policy that this number never decreases. I would be open to suggestions for a different name for the attribute, or adding a comment explaining its purpose.

Comment (dhslim) [reply python_ta/cfg/graph.py]: I see. I thought `block_count` meant the total number of blocks in the graph. I will make a PR with the comment explaining its purpose for now.

**Pull request merged**